### ISSUE TITLE
geographic subject indexing

### ISSUE THREAD

**Issue by ebenenglish** [boston-library/curator#40]
Solr indexing for `DigitalObject` and `Institution` objects requires indexing of complex geographic data that depends on information from [TGN](http://vocab.getty.edu/tgn/) and [Geonames](https://www.geonames.org/) APIs.

### Background
In MODS XML, we have been storing the full object hierarchy, like so:
```xml
<mods:subject authority='tgn' authorityURI='http://vocab.getty.edu/tgn/'
              valueURI='http://vocab.getty.edu/tgn/7008038'>
  <mods:hierarchicalGeographic>
    <mods:region>Ile-de-France</mods:region>
    <mods:country>France</mods:country>
    <mods:continent>Europe</mods:continent>
    <mods:city>Paris</mods:city>
  </mods:hierarchicalGeographic>
  <mods:cartographics>
    <mods:coordinates>48.866667,2.333333</mods:coordinates>
  </mods:cartographics>
</mods:subject>
```
Currenty in Digital Commonwealth, this value is displayed like so:
![DC2_geo-data-display-screenshot](https://user-images.githubusercontent.com/2079425/71998408-912f9600-320d-11ea-8151-9be85f4a4b71.png)

Which has some issues (precision):
* each location is individually linked, so clicking on 'Paris' brings up results from any location with the same name (e.g. Paris, TX)
* hierarchical display may be overkill or not be useful

But our approach also has some benefits (recall):
* Any search (or facet click) for `France` will bring up all records with `Paris`
* Any search (or facet click) for `United States` will bring up all records with any U.S. locale
* Any search (or facet click) for `Suffolk` will bring up all records with any location with that county

------------------------------------
### Changes in New System
Our new data model stores only the following data for geographic subjects:

attribute | example value
-----------|-----------------
label | Paris
area_type | city
authority | AUTHORITY_ID
id_from_auth | 7008038
coordinates | 48.866667,2.333333
bounding box | -93.48 41.93 -86.58 47.37

------------------------------------
### Approach going forward

An alternate approach would be to display locations like so:
![DC3_geo-data-display-mock](https://user-images.githubusercontent.com/2079425/71998950-832e4500-320e-11ea-887e-19933b74e46e.png)

The link could search for `Paris, France`, or both 'Paris' and 'France' (depending on how we decide to index the data). This would potentially be simpler, more in line with how users understand/read locations, and improve search precision.

**QUESTION 1:** 
Should we:
- Adopt this new approach to displaying geographic data?
- Keep the existing display
- Do something else entirely?

**QUESTION 2:** 
How do we want to approach faceting? Should facet values be:
```
Athol, Massachusetts
Paris, France
```
or
```
Athol
France
Massachussetts
Paris
```
**QUESTION 3:** 
Do we continue to index the full geographic hierarchy? E.g.:
* add county for each municipality
* add country for each location

------------------------------------

### Solr indexing details stuff (WIP)

To support our approach, Solr should have the following fields:

field name | example value | use case
--------------|----------------------|------------
subject_geographic_tim | Paris | Text for search. All term labels and parents added to this field
subject_geographic_ssim | Paris | Facet values. All term labels and parents added to this field
subject_geo_label_sim | Paris | used to refine MLT results. add subject label for TGN/Geonames subjects
subject_coordinates_geospatial | 48.866667,2.333333 | for coordinate searches
subject_point_geospatial | 48.866667,2.333333 | not used, but could be useful if we want to only search coordinate points
subject_bbox_geospatial | ENVELOPE(2.06, 2.54, 49.0, 48.74) | not used, but could be useful if we want to only search bounding boxes
subject_geojson_facet_ssim | [{"type":"Feature","geometry":{"type":"Point","coordinates":[2.333333,48.866667]},"properties":{"placename":"Paris, France"}}, "{"type":"Feature","geometry":{"type":"Polygon","coordinates":[[[2.06,48.74],[2.54,48.74],[2.54,49.0],[2.06,49.0],[2.06,48.74]]]},"bbox":[2.06,48.74,2.54,49.0]}"] | for map display
subject_hiergeo_geojson_ssm | {"type":"Feature","geometry":{"type":"Point","coordinates":[2.333333,48.866667]},"properties":{"country":"France","region":"Ile-de-France","city":"Paris"}} | For display of TGN/hierarchical subjects
subject_geo_nonhier_ssm | 13 Village Way | For display of Geonames, uncontrolled, or other non-TGN subjects
subject_geo_country_ssim | France  | not currently used, but could be useful for future mapping/visualization
subject_geo_state_ssim | Texas  | not currently used, but could be useful for future mapping/visualization
subject_geo_county_ssim | Suffolk | not currently used, but could be useful for future mapping/visualization
subject_geo_city_ssim | Boston | not currently used, but could be useful for future mapping/visualization
subject_geo_citysection_ssim | Jamaica Plain | not currently used, but could be useful for future mapping/visualization

These fields are DEPRECATED:

field name | example value | rationale
--------------|----------------------|------------
subject_geo_province_ssim | Quebec | we've never used this, values already exist in `subject_geographic_ssim`
subject_geo_region_ssim | Ile-de-France | we've never used this, values already exist in `subject_geographic_ssim`
subject_geo_territory_ssim | Puerto Rico  | we've never used this, values already exist in `subject_geographic_ssim`
subject_geo_island_ssim | Chebeague Island | we've never used this, values already exist in `subject_geographic_ssim`
subject_geo_area_ssim | New England | we've never used this, values already exist in `subject_geographic_ssim`

**Comment by ebenenglish** [comment]
Some initial work done on `geo-subject-indexing` branch.

**Comment by aclooney** [comment]
@ebenenglish 
QUESTION 1:
- Heirarchical display is useful for the region name -- not so much with a city like Paris. But in Massachusetts, listing the county is very useful for archival/genealogy research. 
As in: Massachusetts > Hampshire County > Northampton > Leeds
- I like the simplicity of just saying Hadley, Massachusetts. Is there an option to display just that by default, but an optional display of all the geographic info available so people can see county or village names if they want? 


QUESTION 2:
- Heirarchical is better for this, but I'm not sure how to achieve that. Would be interested to see display options. 

QUESTION 3: See what I said above. If we can make a simple display with the option of a more complicated one, that would be best.

**Comment by ebenenglish** [comment]
Closed via #77
